datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -351,41 +362,62 @@ pub async fn from_substrait_rel(
                 None => None,
             };
             // If join expression exists, parse the `on` condition expression, build join and return
-            // Otherwise, build join with koin filter, without join keys
+            // Otherwise, build join with only the filter, without join keys
             match &join.expression.as_ref() {
                 Some(expr) => {
                     let on =
                         from_substrait_rex(expr, &in_join_schema, extensions).await?;
                     let predicates = split_conjunction(&on);
                     // TODO: collect only one null_eq_null
-                    let join_exprs: Vec<(Column, Column, bool)> = predicates
-                        .iter()
-                        .map(|p| match p {
+                    // The predicates can contain both equal and non-equal ops.
```

Comment:
I'm not sure if this is a good idea, it seems to make things more complex.

Reply:
@Dandandan Thank you for pointing this out. I agree that this makes things more complicated, and my apologies for the inaccurate explanation of the issue. I added a comment to correct my description of the [issue](https://github.com/apache/arrow-datafusion/issues/7611#issuecomment-1729906548).

The high-level idea is that join `filter` and `post_join_filter` do not have the same meaning semantically. The former is for filtering input during/pre join, the latter is for filtering the output of the join (post-join). Currently in `datafusion`, we do not have a field in join relation that represents a post-join predicate (the parser/logical optimizer takes care of creating an appropriate filter relation if necessary). So the producer should only generate plans with `None` as `post_join_filer`.

I'll modify the consumer to throw an error for now if there's a `post_join_filter`. Later, we can wrap the join relation with a filter relation if we want to support `post_join_filter`, or if the later version of `datafusion` supports `post_join_filter` directly in the join relation, then we can also add the support in both producer and consumer as well.

Reply:
I looked into the substrait spec, and it doesn't really talk about what semantics of the `post_join_filter` is: https://substrait.io/relations/logical_relations/#join-operation
https://github.com/search?q=repo%3Asubstrait-io%2Fsubstrait%20post_join_filter&type=code

There is a subtle distinction between non equality filters applied *during* the join (in the `ON` clause) and applied post join for non-INNER joins: for non inner joins the filters during the join don't filter out input rows (they still come out, just with NULL matches)

> So the producer should only generate plans with None as post_join_filer. 

This makes sense to me

For the consumer, there is already code in DataFusion that breaks up an arbitrary `Expr` into equality predicates and others. This is how the SQL frontend creates a Join (a single expr):

https://github.com/apache/arrow-datafusion/blob/a514b6752b063a5a3006aa114297520a933339b0/datafusion/sql/src/relation/join.rs#L134-L141

I think we could do the same here in the subtrait consumer which would be much simpler, and would let the normal DataFusion optimization machinery work.

Reply:
Thanks @alamb for the pointer! I'll look into this and update the consumer.